Add A New Truffle Random Data Source Example (#46)
oraclize/ethereum-examples: solidity/truffle-examples/random-datasource/README.md

# :wrench: :construction: Testing Oraclize's Random Example.

&nbsp;

This repo is to demonstrate how you would set up an Oraclize smart-contract development environment using Truffle to do most of the heavy lifting for you. Head on over to the `./test` folder to examine the javascript files that thoroughly test the smart-contract, which latter you will find in `./contracts`.

## :page_with_curl:  _Instructions_

> Since the random datasource is currently available on the Ethereum mainnet and on all Ethereum public testnets only (Rinkeby, Kovan, Ropsten-revival); and it is *not integrated yet with private blockchains/testrpc/browser-solidity-vmmode*, you will need to **get a mnemonic passphrase and an infura key** to make the example work.

Reviewer: ...and some test net ethereum!

Author: Yeah, for now goerli doesn't

**A)** Go on the infura website to get an infura key: [infura.io](https://infura.io).
The key is a 32 HEX characters string

**B)** Get a mnemonic passphrase, an easy way is to get the Metamask one:
[metamask.io](https://metamask.io/)

> **C)** Create a new file `apikeys.js` or `.env` and *add the mnemonic passphrase and the infura key to it*, such as:

Reviewer: Just adding something to a `.env` file doesn't magically make it an environment variable, and so neither the code I wrote in the `truffle.js` nor that you copy/pasted into utils (they should share it, btw, with proper exporting, rather than copy/pasting) would pick up the keys from a `.env` file if the user chooses this option.

Author: This line `require('dotenv').config()` adds the `.env` file variables to the environment variable stack.

Reviewer: And where is this line? Because as of your latest commit @ time of writing (2ed7a8a) in this PR, neither this line nor the package in the `package.json` exist.

I am aware that that is how you parse `.env` to achieve the environment variable, but I'm pointing out the lack of it in this repo, making the `.env` pointless, and the recommendation for using it in the `README` also pointless.

Author: The line is the fix...

Reviewer: I am well aware, but you stated the above line as if:

a) It's already in the repo and working
& b) I don't know how to use an environment variable in a node project.

But the point I'm making isn't me asking how to do it, it's that you've suggested a user (who is likely to be a beginner given that they are looking at our truffle examples) to use an environment variable in a `.env` file when that won't work in the repo as it stands at the time of reviewing. 

And so I've raised this entire point in the review precisely because it's _not_ in the repo, and so _doesn't_ work.

Author: No, I stated:

this line will now fix the issue. 

If it was not clear the point I'm making with: 

> This line require('dotenv').config() adds the .env file variables to the environment variable stack.

is: there was an issue you pointed out, and I solved it adding that line. I wasn't claiming for obvious reasons that was already in the repo and working.

Reviewer: >This line require('dotenv').config() adds the .env file variables to the environment variable stack.

You statement is declarative and conveys no sense of having made a fix whatsoever, nor intention to do so. It simply points out what we all know about how to use an environment variable in a node project.

This review is me pointing out things that need fixing.

Author: I am confused. Again, what I want to say here is that editing the file adding `require('dotenv').config()` fix now the issue you pointed out.

```javascript
// apikeys.js example

> module.exports = {

Reviewer: Again, you suggest this as a format for both `apikeys` and `.env`, but for the latter this will not work.

Author: I was suggesting this just for apikeys

Reviewer: In which case where is the bit about how to use the `.env` file for the reader of the tutorial?

Author: there is no reference.

Reviewer: I know, do you see what I'm getting at...?

Author: What's going on here? Do you want me to add more details to the doc?

```
**C)** Create a new file `apikeys.js` or `.env` and *add the mnemonic passphrase and the infura key to it*
```

with the new `require('dotenv').config()` adding variables to `.env` will be enough; but, if it's not so clear, I will add an example like this in the doc:

```
// .env example 

mnemonic = 'word1 word2 ...'
infuraKey = 'f09924890213480...'
```

Reviewer: Yes, that's what's going on. If you're mentioning `.env` in the `README`, those `.env` need to work. And currently, as of the reviewed commit 2ed7a8a, they do not.

Author: .env details added to the readme

  mnemonic: 'word1 word2 word3 word4 word5 word6 word7 word8 word9 word10 word11 word12'
  infuraKey: '0123456789abcdef0123456789abcdef'
}
```

> Please note that those must be kept secure and the one provided here are just to help the user with the tutorial.

Reviewer: ```suggestion
Please note that these keys must be kept secure. The ones provided here are just placeholders to help with the tutorial.
```

---

Then you can proceed with the following console instructions:

> **1)** Fire up your favourite console & make sure you have Truffle globally 5 installed:

Reviewer: The following instructions all need changing to be brought in line with the other trufle examples, which now use `npx` instead of global installs.

# :wrench: :construction: Testing Oraclize's Random Example.

&nbsp;

This repo is to demonstrate how you would set up an Oraclize smart-contract development environment using Truffle to do most of the heavy lifting for you. Head on over to the `./test` folder to examine the javascript files that thoroughly test the smart-contract, which latter you will find in `./contracts`.

## :page_with_curl:  _Instructions_

> The random datasource is currently available on the Ethereum mainnet and on some Ethereum public testnets only (Rinkeby, Kovan, and Ropsten Revival); and it is *not integrated yet with private blockchains/testrpc/browser-solidity-vmmode*, so you will need to **get a mnemonic passphrase and an infura key** to make the example work.

Reviewer: Please remove the bold & italic formatting. Mixing the two and overusing them both looks bad and makes comrehension harder.

**1)** Go on the infura website to get an infura key: [infura.io](https://infura.io).

> The key is a 32 HEX characters string

Reviewer: No new line.

> **2)** Get a mnemonic passphrase, an easy way is to get the Metamask one:

Reviewer: ~~the~~ a Metamask one

> [metamask.io](https://metamask.io/)

Reviewer: No newline.

**3)** Fire up your favourite console & clone this repo somewhere:

__`❍ git clone https://github.com/oraclize/ethereum-examples.git`__

**4)** Enter this directory & install dependencies:

__`❍ cd ethereum-examples/solidity/truffle-examples/random-datasource && npm install`__

> **5)** In `ethereum-examples/solidity/truffle-examples/random-datasource`, create a new file `apikeys.js` or `.env` and *add the mnemonic passphrase and the infura key to it*, such as:

Reviewer: Rm italics.

```javascript
// apikeys.js example

module.exports = {
  mnemonic: 'word1 word2 word3 word4 word5 word6 word7 word8 word9 word10 word11 word12',
  infuraKey: '0123456789abcdef0123456789abcdef'
}
```

or

```
// .env example

mnemonic = 'word1 word2 word3 word4 word5 word6 word7 word8 word9 word10 word11 word12'
infuraKey = '0123456789abcdef0123456789abcdef'
```

**6)** Compile with Truffle:

__`❍ npx truffle compile`__

**7)** Test the contract with Truffle leveraging a testnet, such as Rinkeby:

__`❍ npx truffle test --network rinkeby`__

&nbsp;

## :camera: Passing Tests:

[The passing tests!](random-datasource-test.jpg)

&nbsp;

## :black_nib: Notes:

> __❍___ Note that the Infura key and the Metamask passphrase must be kept secure. The ones provided in the readme are just placeholders to help the user with the tutorial.

Reviewer: You've snuck an extra `_` here.

```suggestion
__❍__ Note that your own Infura key and the Metamask passphrase must be kept secure. The ones provided here are simply placeholders.
```

# :wrench: :construction: Testing Oraclize's Random Example.

&nbsp;

This repo is to demonstrate how you would set up an Oraclize smart-contract development environment using Truffle to do most of the heavy lifting for you. Head on over to the `./test` folder to examine the javascript files that thoroughly test the smart-contract, which latter you will find in `./contracts`.

## :page_with_curl:  _Instructions_

Since the random datasource is currently available on the Ethereum mainnet and on some Ethereum public testnets only (Rinkeby, Kovan, and Ropsten Revival); and it is *not integrated yet with private blockchains/testrpc/browser-solidity-vmmode*, you will need to **get a mnemonic passphrase and an infura key** to make the example work.

**A)** Go on the infura website to get an infura key: [infura.io](https://infura.io).
The key is a 32 HEX characters string

**B)** Get a mnemonic passphrase, an easy way is to get the Metamask one:
[metamask.io](https://metamask.io/)

> **C)** In `ethereum-examples/solidity/truffle-examples/random-datasource`, create a new file `apikeys.js` or `.env` and *add the mnemonic passphrase and the infura key to it*, such as:

Reviewer: These instructions come _before_ the main instructions, which latter's first command involves cloning this repo. And so it's confusing, since this earlier line assumes the user has _already_ cloned the repo. 

It's probably best to merge to two instruction sets into one.
